### ISSUE TITLE
Future Features for Raid Emulator

### ISSUE THREAD

**Issue by valarnin** [quisquous/cactbot#3275]
Now that the raid emulator typescript conversion is complete, I wanted to solicit input on future features/ideas for the emulator.

I'll maintain a task list here with features/status/requester.

- [ ] N/A - Translation support - WIP
- [ ] valarnin - Map/position display during playback - Not started yet
- [ ] valarnin - Shield display on party info - Not sure if possible
- [ ] quisquous - Encounter analysis for testing purposes - Not started yet
  - quisquous suggested this shortly before the typescript conversion process started, being able to pass logs through the raid emulator analysis process to get information on executed triggers for testing, IIRC?

**Comment by quisquous** [comment]
Hmm.  I feel like my opinions of what the raid emulator should be for diverge from what you think it should be.

For me, my view of the goal of raid emulator is to test raidboss triggers, in other words testing what raidboss would have emitted during a fight.  My uses of the tool are:
* 98% reading the analysis of trigger output for a particular player (or console.log lines I've added to debug)
* 2% real time playback (ultimates + one fight an expansion) to verify tricky timings / colliding outputs

(For what it's worth, I would like to add oopsy emulation at some point, but I don't think it needs the same real time nature that raidboss emulation does and was planning to do it in a "headless" way that just generated the events from a log.  I don't think it would make sense to add it to raidemulator.)

I think I've mentioned this before, but I still don't find the hp/mp bars useful.  I know you said it helped you track down a mitigation issue, but I would strongly argue that you can do that from looking at video, fflogs, or just the logs themselves more efficiently in most cases.  Further things in the "replayer" space like shielding, maps, positioning, etc are in this similar "neat idea but I have a hard time seeing the goal" category.  In particular, I think there are already tools that do this really well like the fflogs replayer, which also has lb, dps meters, effects, cast bars.  See: [example](https://www.fflogs.com/reports/F42Yg8tkmapKPhMX#fight=22&view=replay).  What problem are you trying to solve that requires recreating that entire tool?  Sorry, I'm not trying to be argumentative here, but I just feel like there's a very large disconnect between our goals here and I have a hard time understanding what you're aiming to achieve.

I get that raidemulator has a bunch of log loading and parsing built in, but other than that, I think the "replayer" vs "emulator" use cases are very distinct from each other.  If this is something you'd really like to do, would it make sense to separate out the "log parsing backend" in such a way that you could have multiple frontends, e.g. a simpler emulator on its own and a simpler replayer on its own and each could be simpler/faster/less entangled with the needs of the other?

**Comment by quisquous** [comment]
> quisquous suggested this shortly before the typescript conversion process started, being able to pass logs through the raid emulator analysis process to get information on executed triggers for testing, IIRC?

Yeah, I would like to be able to run the emulator analysis in some sort of "headless" fashion, i.e. log lines + player as inputs, and trigger events (time, output of all the fields on the trigger) as outputs to then feed into some sort of testing.

**Comment by quisquous** [comment]
I left a long comment and github looked like it double posted it, so I deleted one, and it deleted the other.  Excellent.  /o\

**Comment by quisquous** [comment]
Maybe if you or somebody else got this deleted comment by email you can just paste it so I don't have to reformulate all those thoughts.  (Sorry!!)

**Comment by Akurosia** [comment]
```Hmm. I feel like my opinions of what the raid emulator should be for diverge from what you think it should be.
For me, my view of the goal of raid emulator is to test raidboss triggers, in other words testing what raidboss would have emitted during a fight. My uses of the tool are:
•	98% reading the analysis of trigger output for a particular player (or console.log lines I've added to debug)
•	2% real time playback (ultimates + one fight an expansion) to verify tricky timings / colliding outputs
(For what it's worth, I would like to add oopsy emulation at some point, but I don't think it needs the same real time nature that raidboss emulation does and was planning to do it in a "headless" way that just generated the events from a log. I don't think it would make sense to add it to raidemulator.)
I think I've mentioned this before, but I still don't find the hp/mp bars useful. I know you said it helped you track down a mitigation issue, but I would strongly argue that you can do that from looking at video, fflogs, or just the logs themselves more efficiently in most cases. Further things in the "replayer" space like shielding, maps, positioning, etc are in this similar "neat idea but I have a hard time seeing the goal" category. In particular, I think there are already tools that do this really well like the fflogs replayer, which also has lb, dps meters, effects, cast bars. See: example. What problem are you trying to solve that requires recreating that entire tool? Sorry, I'm not trying to be argumentative here, but I just feel like there's a very large disconnect between our goals here and I have a hard time understanding what you're aiming to achieve.
I get that raidemulator has a bunch of log loading and parsing built in, but other than that, I think the "replayer" vs "emulator" use cases are very distinct from each other. If this is something you'd really like to do, would it make sense to separate out the "log parsing backend" in such a way that you could have multiple frontends, e.g. a simpler emulator on its own and a simpler replayer on its own and each could be simpler/faster/less entangled with the needs of the other?
```

Thats what in my emails :)

**Comment by valarnin** [comment]
> For me, my view of the goal of raid emulator is to test raidboss triggers, in other words testing what raidboss would have emitted during a fight. My uses of the tool are:
> - 98% reading the analysis of trigger output for a particular player (or console.log lines I've added to debug)
> - 2% real time playback (ultimates + one fight an expansion) to verify tricky timings / colliding outputs
> 
> (For what it's worth, I would like to add oopsy emulation at some point, but I don't think it needs the same real time nature that raidboss emulation does and was planning to do it in a "headless" way that just generated the events from a log. I don't think it would make sense to add it to raidemulator.)
> I think I've mentioned this before, but I still don't find the hp/mp bars useful. I know you said it helped you track down a mitigation issue, but I would strongly argue that you can do that from looking at video, fflogs, or just the logs themselves more efficiently in most cases. Further things in the "replayer" space like shielding, maps, positioning, etc are in this similar "neat idea but I have a hard time seeing the goal" category. In particular, I think there are already tools that do this really well like the fflogs replayer, which also has lb, dps meters, effects, cast bars. See: example. What problem are you trying to solve that requires recreating that entire tool? Sorry, I'm not trying to be argumentative here, but I just feel like there's a very large disconnect between our goals here and I have a hard time understanding what you're aiming to achieve.

In the case of HP/MP it's more a case of "we're already tracking this information for `CombatantTracker` so that `getCombatants` can work, so why not display it?"

For maps and positioning, it can help with setting up/debugging triggers that call to `getCombatants` without having to sit down and do the math yourself, which is why I wanted to add it.

Shields were more of a case of "if we've got the info and we're already showing the other stuff, why not show shields too if they're available". I'm not really tied to the idea there and if it turns out shields aren't easily available right now, that's fine too.

> I get that raidemulator has a bunch of log loading and parsing built in, but other than that, I think the "replayer" vs "emulator" use cases are very distinct from each other. If this is something you'd really like to do, would it make sense to separate out the "log parsing backend" in such a way that you could have multiple frontends, e.g. a simpler emulator on its own and a simpler replayer on its own and each could be simpler/faster/less entangled with the needs of the other?

It's perfectly viable to make a separate tool available that uses the same backend but only shows the info in the trigger analysis column. That tool would be a bit faster but not significantly, since most of the slowdown right now is in running the log lines through PopupText 8 times (once for each player).

There are some potential performance gains there by flagging lines that actually match trigger regexes on the first player's analysis and only processing those lines on future analysis but then there's potential for delayed triggers and such to drift, so making that code faster would be a bit complex.

Realistically, for any alternative tool, all you really need to do is something like:

```ts
const options = {}; // Pull raidboss options as normal
const log = ''; // full log read in from somewhere
const logLines = log.split(/\r?\n/gm);
const encDay = ''; // figure out from lines or use a dummy value
const encZoneId = ''; // figure out from lines or use a dummy value
const encZoneName = ''; // figure out from lines or use a dummy value
const enc = new Encounter(encDay, encZoneId, encZoneName, logLines);
enc.initialize();
// AnalyzedEncounter only needs emulator for one thing, can be refactored
const emulator = new RaidEmulator(options);
const analyzedEnc = new AnalyzedEncounter(options, enc, emulator);
await analyzedEnc.analyze();
// At this point you can access analyzedEnc.perspectives to get executed trigger data
```

**Comment by quisquous** [comment]
> Thats what in my emails :)

(Thank you!  So embarrassing to write a big comment and then accidentally delete it.  WHOOPS.)

> In the case of HP/MP it's more a case of "we're already tracking this information for `CombatantTracker` so that `getCombatants` can work, so why not display it?"

Specifically, in the case of HP/MP, it's being shown at the expense of space to show the name or more combatants for alliance raids.  I have to click through all of the names to find the right player whose perspective I'm trying to debug.

More generally, it's an argument for simplicity.  Making raidboss emulator be both a raidboss debugger and an fflogs replayer makes them both necessarily more complicated.  Features get entangled; files and CSS gets more complicated; performance is harder to evaluate; it's harder to refactor and understand; bugs in lower priority features can break higher priority features.

It feels like antipattern to me to have one tool try to do too many things.  Ultimately, I think there's a disconnect between our design goals here.  The answer to "why not display it" to me is because "that's not in line with what I consider to be the goal of raid emulator, so don't add unnecessary complexity".

> For maps and positioning, it can help with setting up/debugging triggers that call to `getCombatants` without having to sit down and do the math yourself, which is why I wanted to add it.

Maybe it's just my workflow, I don't think there's been a case where I've wanted a map to debug these.  Most of the time it's fairly clear to me that "the mob in the east facing south" from looking at the log line itself or from a console log message.  The hard part for me is making sure that https://github.com/quisquous/cactbot/blob/main/ui/raidboss/data/05-shb/eureka/delubrum_reginae_savage.ts#L230 calls the right thing in that case, and a map isn't going to solve that.

It's also still not clear to me what this is solving that the fflogs replayer isn't already solving.  I'm not trying to be rude here, but when there's already an existing tool that solves a problem then I always want to understand what the goals are in reimplementing it.

> > I get that raidemulator has a bunch of log loading and parsing built in, but other than that, I think the "replayer" vs "emulator" use cases are very distinct from each other. If this is something you'd really like to do, would it make sense to separate out the "log parsing backend" in such a way that you could have multiple frontends, e.g. a simpler emulator on its own and a simpler replayer on its own and each could be simpler/faster/less entangled with the needs of the other?
> 
> It's perfectly viable to make a separate tool available that uses the same backend but only shows the info in the trigger analysis column. That tool would be a bit faster but not significantly, since most of the slowdown right now is in running the log lines through PopupText 8 times (once for each player).

Unrelatedly, maybe we shouldn't do that? If that's the biggest slowdown, then maybe we should skip that feature so that everything loads more quickly.  On a similar note, I know you've optimized `Data` copying because it's slow, and I don't know that I use that either.  Would it be speed things up to not store it?

**Comment by valarnin** [comment]
I think that maybe I didn't explain clearly enough?

Functionally, right now it's running every log line of an encounter against every trigger loaded for every player.

For instance, e9s loads 42 `netTriggers`, which have to run against 8 players, for every log line in the encounter (36180 in my test encounter), totalling 12,156,480 regex tests.

Running a performance analysis shows the following usage:
![image](https://user-images.githubusercontent.com/6119598/127080723-9d3bbd25-81f4-421b-adeb-fd885312b1a7.png)

For reference, in this same perf analysis, the data portion is relatively speaking tiny:

![image](https://user-images.githubusercontent.com/6119598/127080871-e5e7f72a-12de-4592-9ea2-b53bc2b259af.png)

The performance gain here would be to only analyze the first player immediately, and analyze the rest of the players in a worker, populating their data as it becomes available. However, I'm not sure how exactly workers would pan out with any sort of command-line node execution.

There's additional performance gains that could be gotten by adding the same sort of logic that's in the parser plugin now for self/party/alliance as a config setting, but we'll have to carry forward the 0x02 (player change) and 0x0B (party change) lines for that to be viable. Part of the extreme time it takes to load the DRS Trinity Avowed encounter from that issue is because it's running 209,059,536 regex tests.

Finally, there's potentially only testing lines that can match a `NetRegexes` or `Regexes` entry. Here's the count of line events for that DRS Trinity Avowed encounter:

```
0: 12060
3: 168
4: 136
12: 3
20: 439
21: 12048
22: 2457
23: 14
24: 4332
25: 25
26: 8017
27: 3
28: 16
29: 5
30: 4586
31: 97
33: 12
34: 28
36: 172
37: 15550
38: 11661
39: 8454
251: 1
```

If you want to check the amount of regex tests ran yourself, you can run this js with any loaded encounter:

```ts
window.raidEmulator.emulator.currentEncounter.encounter.logLines.length * window.raidEmulator.emulator.currentEncounter.encounter.combatantTracker.partyMembers.length * (window.raidEmulator.popupText.netTriggers.length + window.raidEmulator.popupText.triggers.length)
```

**Comment by quisquous** [comment]
> The performance gain here would be to only analyze the first player immediately, and analyze the rest of the players in a worker, populating their data as it becomes available. However, I'm not sure how exactly workers would pan out with any sort of command-line node execution.

Yeah.  I think if you only analyzed the first player *and* saved which player was active (and not just which fight was active), then it would save a lot of time when reloading.  (Making code changes, auto-reloading, and looking at how the analysis/console.log output changes is probably my most common debugging workflow.)

Re: command-line node execution.  I think that's fine, because I think any execution of workers would be at a higher level than what the command-line execution should be, in my opinion?

My hope is that there could be some single API that takes as input a set of network log lines, an array of triggersets, and options (like current player name).  And as output, it has some array of actions that were emitted.  Probably something close to what AnalyzedEncounter is doing.  The use case for me is that it might be nice to have a command-line utility similar to test timeline where I can print what triggers occur (for a particular player, for a particular encounter).  The other use case is to turn that into some sort of regression testing so that I can have a bunch of test cases with log lines and verify that things like TEA nisi or other complicated triggers are working (and keep working).

**Comment by valarnin** [comment]
Here's an example command-line tool that "should" work for running a file through and getting executed triggers for a player for every encounter in the file.

I say "should" because I wasn't able to get command-line node to load the raidboss manifest file, I just don't have the requisite knowledge of that portion of the project to get it to work.

I'd like to drop the requirement for the `RaidEmulator` class but it's currently required for the timeline trigger support and I need to do a bit of refactoring to fix that (it's not really used during analysis but I didn't subclass separately to allow it to be left out).

```ts
#!/usr/bin/env ts-node

import { readFileSync } from 'fs';
import { argv, exit } from 'process';

import { isLang } from '../resources/languages';
import AnalyzedEncounter from '../ui/raidboss/emulator/data/AnalyzedEncounter';
import Encounter from '../ui/raidboss/emulator/data/Encounter';
import LogEventHandler from '../ui/raidboss/emulator/data/LogEventHandler';
import LineEvent from '../ui/raidboss/emulator/data/network_log_converter/LineEvent';
import NetworkLogConverter from '../ui/raidboss/emulator/data/NetworkLogConverter';
import RaidEmulator from '../ui/raidboss/emulator/data/RaidEmulator';
import EmulatorCommon from '../ui/raidboss/emulator/EmulatorCommon';
import defaultOptions from '../ui/raidboss/raidboss_options';

const file = argv[1];
const player = argv[2];

if (!file || !player)
  exit(1);

const converter = new NetworkLogConverter();
const localLogHandler = new LogEventHandler();
const options = { ...defaultOptions };
const localLang = Intl.DateTimeFormat().resolvedOptions().locale;
options.ParserLanguage = isLang(localLang) ? localLang : 'en';
options.DisplayLanguage = isLang(localLang) ? localLang : 'en';
const emulator = new RaidEmulator(options);

const fileBuf = readFileSync(file, { encoding: 'utf8' });

const encs: AnalyzedEncounter[] = [];

localLogHandler.on('fight', (day: string, zoneId: string, zoneName: string, lines: LineEvent[]) => {
  const enc = new Encounter(day, zoneId, zoneName, lines);
  enc.initialize();
  if (enc.shouldPersistFight())
    encs.push(new AnalyzedEncounter(options, enc, emulator));
});

const lines = converter.convertFile(fileBuf);

localLogHandler.parseLogs(lines);

void (async () => {
  for (const enc of encs) {
    await enc.analyzeFor(player).then(() => {
      const triggers = enc.perspectives[player]?.triggers;
      const playerName = enc.encounter.combatantTracker?.combatants[player]?.name;
      const encString = `${new Date(enc.encounter.initialTimestamp).toUTCString()} - ${enc.encounter.encounterZoneName} - ${enc.encounter.combatantTracker?.getMainCombatantName() ?? ''}`;
      if (!triggers) {
        console.log(`No triggers fired for player ${playerName ?? ''} (${player}) in encounter ${encString}`);
      } else {
        console.log(`${encString}`);
        for (const trigger of triggers) {
          let triggerString = EmulatorCommon.timeToString(trigger.resolvedOffset, true);
          triggerString += ` - ${trigger.status.executed ? 'Executed    ' : 'Not Executed'}`;
          triggerString += ` - ${trigger.triggerHelper.trigger.id ?? 'NO ID'}`;
          triggerString += ` - ${trigger.status.responseType ?? 'NO TYPE'}`;
          triggerString += ` - ${trigger.status.responseLabel ?? 'NO LABEL'}`;
          console.log(triggerString);
        }
      }
    });
  }
})();
```
